### DISCUSSION TITLE
[Question about native contacts analysis] Can I only compute heavy atom pairs that their residue number is not same?

### DISCUSSION THREAD

**Discussion by narubi2** [MDAnalysis/mdanalysis#4498] [General]
Hi,

I'm working with MDAnalysis and have a question about 'native contact analysis'
I'm trying to get Q value (native contact value) of all heavy atoms.

I tried as follows:

```python
# seletion
sel_ca_atom1 = "protein and not (name H*)"
sel_ca_atom2 = "protein and not (name H*)"
ref_ca_atom1 = u.select_atoms(sel_ca_atom1)
ref_ca_atom2 = u.select_atoms(sel_ca_atom2)

# calculate
ca = contacts.Contacts(u, select=(sel_ca_atom1, sel_ca_atom2),
                        refgroup=(ref_ca_atom1, ref_ca_atom2),
                        radius=4.5,
                        method='soft_cut',
                        kwargs={'beta': 5.0,
                                'lambda_constant': 1.5}).run()
```

It worked, but it took a long time to compute. Additionally it was hard to interpret the result.
I read the paper about Q value (https://doi.org/10.1073/pnas.1311599110) again, and found they compute only pairs of heavy atoms i and j belonging to residues θi and θj are in contact provided that |θi - θj| > 3

Is there a way to only count pairs of heavy atoms whose their residue number difference is larger than 3?

Any suggestion or advice would be appreciated.
Thanks

**Comment by orbeckst** [comment]
The [current code for Contacts](https://docs.mdanalysis.org/stable/_modules/MDAnalysis/analysis/contacts.html#Contacts) does not allow you to exclude neighbors and nearest neighbors. You could probably write your own version that does something like it. When you look at the source code's `Contact._single_frame` function https://github.com/MDAnalysis/mdanalysis/blob/0582265996b392da382f658b7f0805ca250e1233/package/MDAnalysis/analysis/contacts.py#L486-L499
then you see that we calculate _all_ distances with
```python
     d = distance_array(self.grA.positions, self.grB.positions, 
                         box=self._get_box(self._ts)) 
```
but then filter these distances with `initial_contacts`.

`initial_contacts` is a boolean array of the same shape as `d`. So if you modify `initial_contacts` to exclude the contacts that you don't want to see then the `Contacts` code will produce what you want, e.g., *in the special case that your selections A and B are a linear sequence of residues* then you would set the diagonal and the two off-diagnals to `False` and they should be excluded from the calculations, perhaps something like this
```python
import numpy as np
c = Contacts(...)
initial_contacts = c.initial_contacts[0]  # depending on how you run Contacts, can contain more
mask = np.ones(initial_contacts.shape, dtype=bool)
# keep second, first, and main diagonal at True, set upper and lower triangles to False
mask = np.triu(mask, k=-2) * np.tril(mask, k=2)
# set contacts on 2nd, 1st, and main diagonal to False
initial_contacts[mask] = False
# and update the class before running analysis
c.initial_contacts[0] = initial_contacts

# run analysis
c.run()
```

The above code will not just work — I only had a quick look at the source code and I think you can hack it using the ideas above. Perhaps the notes on [native contact analysis](https://userguide.mdanalysis.org/stable/examples/analysis/distances_and_contacts/contacts_native_fraction.html) are also useful.

You could also rewrite the `Contacts` code itself.

In any case, you'll have some coding to do in order to achieve what you want.

**Reply by narubi2** [reply]
Hi orbeckst.

Thank you for reply
I'll try your suggestions.
